Modules/_copy.c

```diff
@@ -0,0 +1,464 @@
+        Py_SETREF(elem, do_deepcopy(module, elem, memo));
+        if (elem == NULL) {
+            Py_DECREF(y);
+            return NULL;
+        }
+        /*
+         * This really should not happen, but let's just return what's left in
+         * the list.
+         */
+        if (i >= PyList_GET_SIZE(y)) {
+            Py_DECREF(elem);
+            break;
+        }
+        PyList_SetItem(y, i, elem);
+    }
+    return y;
+}
+
+/*
+ * Helpers exploiting ma_version_tag. dict_iter_next returns 0 if the dict is
+ * exhausted, 1 if the next key/value pair was succesfully fetched, -1 if
+ * mutation is detected.
+ */
+struct dict_iter {
+    PyObject* d;
+    uint64_t tag;
+    Py_ssize_t pos;
+};
+
+static void
+dict_iter_init(struct dict_iter* di, PyObject* x)
+{
+    di->d = x;
+    di->tag = ((PyDictObject*)x)->ma_version_tag;
+    di->pos = 0;
+}
+
+static int
+dict_iter_next(struct dict_iter* di, PyObject** key, PyObject** val)
+{
+    int ret = PyDict_Next(di->d, &di->pos, key, val);
+    if (((PyDictObject*)di->d)->ma_version_tag != di->tag) {
+        PyErr_SetString(PyExc_RuntimeError, "dict mutated during iteration");
+        return -1;
+    }
+    return ret;
+}
+
+static PyObject*
+deepcopy_dict(PyObject* module, PyObject* x, PyObject* memo, PyObject* id_x,
+              Py_hash_t hash_id_x)
+{
+    PyObject* y, * key, * val;
+    Py_ssize_t size;
+    int ret;
+    struct dict_iter di;
+
+    assert(PyDict_CheckExact(x));
+    size = PyDict_Size(x);
+
+    y = _PyDict_NewPresized(size);
+    if (y == NULL) {
+        return NULL;
+    }
+    if (_PyDict_SetItem_KnownHash(memo, id_x, y, hash_id_x) < 0) {
+        Py_DECREF(y);
+        return NULL;
+    }
+
+    dict_iter_init(&di, x);
+    while ((ret = dict_iter_next(&di, &key, &val)) > 0) {
+        Py_INCREF(key);
+        Py_INCREF(val);
+
+        Py_SETREF(key, do_deepcopy(module, key, memo));
+        if (key == NULL) {
+            Py_DECREF(val);
+            break;
+        }
+        Py_SETREF(val, do_deepcopy(module, val, memo));
+        if (val == NULL) {
+            Py_DECREF(key);
+            break;
+        }
+
+        ret = PyDict_SetItem(y, key, val);
+        Py_DECREF(key);
+        Py_DECREF(val);
+        if (ret < 0) {
+            /* Shouldn't happen - y is presized */
+            break;
+        }
+    }
+    /*
+     * We're only ok if the iteration ended with ret == 0; otherwise we've
+     * either detected mutation, or encountered an error during deepcopying or
+     * setting a key/value pair in y.
+     */
+    if (ret != 0) {
+        Py_DECREF(y);
+        return NULL;
+    }
+    return y;
+}
+
+static PyObject*
+deepcopy_tuple(PyObject* module, PyObject* x, PyObject* memo, PyObject* id_x,
+               Py_hash_t hash_id_x)
+{
+    PyObject* y, * z;
+    int all_identical = 1; /* are all members their own deepcopy? */
+
+    assert(PyTuple_CheckExact(x));
+    Py_ssize_t size = PyTuple_GET_SIZE(x);
+
+    y = PyTuple_New(size);
```

Comment:
With https://github.com/python/cpython/pull/107183 we could use `_PyTuple_NewNoTrack` and add tracking after the tuple is properly initialized.

Reply:
Did you consider `_PyTuple_FromArray`?